core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -0,0 +1,675 @@
+  }
+
+  private boolean commit(Substream winningSubstream) {
+    delayedClientTransport.removeUncommittedRetriableStream(this);
+
+    Collection<Substream> substreamsCopy;
+
+    synchronized (lock) {
+      if (!committed) {
+        committed = true;
+        substreamsCopy = substreams;
+        substreams = Collections.<Substream>singleton(winningSubstream);
+        sharedBuffer = null;
+      } else {
+        return false;
+      }
+    }
+
+    // For hedging only, not needed for normal retry
+    // TODO(zdapeng): also cancel all the scheduled hedges.
+    for (Substream substream : substreamsCopy) {
+      if (substream != winningSubstream) {
+        substream.cancel(Status.CANCELLED);
+      }
+    }
+
+    return true;
+  }
+
+  private void retry() {
+    Substream substream;
+
+    synchronized (lock) {
+      substream = new Substream(); // TODO(zdapeng): update "grpc-retry-attempts" header
+      substreams = Collections.<Substream>singleton(substream);
+    }
+
+    substream.drain();
+  }
+
+  /** Starts the first PRC attempt. */
+  @Override
+  public void start(ClientStreamListener listener) {
+    masterListener = listener;
+    delayedClientTransport.addUncommittedRetriableStream(this);
+    Substream sub;
+
+    synchronized (lock) {
+      // This is the first attempt.
+      sub = new Substream();
+      substreams = Collections.<Substream>singleton(sub);
+    }
+
+    sub.drain();
+    sub.start(new Sublistener(sub));
+
+    // TODO(zdapeng): schedule hedging if needed
+  }
+
+  @Override
+  public void cancel(Status reason) {
+    delayedClientTransport.removeUncommittedRetriableStream(this);
+
+    Collection<Substream> substreamsCopy;
+    synchronized (lock) {
+      substreamsCopy = substreams;
+    }
+
+    for (ClientStream substream : substreamsCopy) {
+      substream.cancel(reason);
+    }
+
+    // TODO(zdapeng): also cancel all the scheduled attempts.
+  }
+
+  /**
+   * Do not use it directly. Use {@link #sendMessage(ReqT)} instead because we don't use InputStream
+   * for buffering.
+   */
+  @Override
+  public void writeMessage(InputStream message) {
+    throw new IllegalStateException("RetriableStream.writeMessage() should not be called directly");
+  }
+
+  /**
+   * Sends out the message if committed, or buffers it first and then sends it out on the current
+   * substream.
+   */
+  void sendMessage(ReqT message) {
+    synchronized (lock) {
+      if (!committed) {
+        sharedBuffer.add(new SendMessageEntry(message));
+      }
+      for (Substream substream : substreams) {
+        substream.sendmessage(message);
+      }
+    }
+  }
+
+  /** Requests if committed, or buffers it first and then requests on the current substream. */
+  @Override
+  public void request(int numMessages) {
+    synchronized (lock) {
+      if (!committed) {
+        sharedBuffer.add(new RequestEntry(numMessages));
+      }
+      for (Substream substream : substreams) {
```

Comment:
I don't think we should be calling the substream with the lock held, as it's asking for a deadlock. We should discuss the locking probably in person.

Reply:
For normal retry, it can be pulled out of the sync block easily by saving a copy ref of the collection in the sync block. For hedging, there are some annoying race conditions.

```
  public void request(int numMessages) {
    // Works for non-hedging case

    Collection<Substream> substreamsCopy;

    synchronized (lock) {
      substreamsCopy = substreams;
      if (!committed) {
        sharedBuffer.add(new RequestEntry(numMessages));
      }
    }

    for (Substream substream : substreamsCopy) {
      substream.request(numMessages);
    }
  }
```

Reply:
uh... I take it back. It is subtle even in normal retry case.

Reply:
Maybe run `substream.delayOrExecute(bufferEntry)` directly and check the tail of the buffer can solve the problem?

```java
  public void request(int numMessages) {
    BufferEntry bufferEntry = new RequestEntry(numMessages);
    Collection<Substream> substreamsCopy;

    synchronized (lock) {
      substreamsCopy = substreams;
      if (!committed) {
        sharedBuffer.add(bufferEntry);
      }
    }

    for (Substream substream : substreamsCopy) {
      substream.delayOrExecute(bufferEntry);
    }
  }
``` 

```java
    void delayOrExecute(BufferEntry bufferEntry) {
      synchronized (lock) {
        if (!upToDate) { // Draining is going on.
          if (committed) {
            buffer.add(bufferEntry);
          }
          // else bufferEntry is added in the sharedBuffer already
          return;
        }
        if (buffer.tail() == bufferEntry) {
          return;
        }
      }
      bufferEntry.runWith(stream);
    }
```